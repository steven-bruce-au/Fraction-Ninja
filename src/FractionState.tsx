
/** Type used to by the FractionComparator to describe the state of a fraction column
*/
export class FractionState {
  numerator: number
  denominator: number
  charts: number // number of whole charts required to render this fraction 
  value: number 

  constructor(numerator: number=1, denominator: number=1) {
  	this.numerator = numerator;
  	this.denominator = denominator;
    this.value = numerator / denominator;
    this.charts = Math.ceil(numerator / denominator);
  }

}