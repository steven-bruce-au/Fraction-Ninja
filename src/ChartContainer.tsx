import React from 'react';
import memoize from "fast-memoize";

import PieFraction from './PieFraction';
import { FractionState } from './FractionState';
import styles from './ChartContainer.module.scss'

type ChartData = {
  equal: boolean,
  fractionState: FractionState,
  maxCharts: number
}

/** Chart container renders multiple PieFractions determined by the ChartData */
export default class ChartContainer extends React.Component<ChartData, {}> {

  constructor(props: ChartData) {
    super(props);
    this.state = {
    };
  }

  chartWidth = memoize(
    (charts: number, maxCharts: number)  =>  { return { width: `${charts * 100/maxCharts}%`}; }
  );

  render() : JSX.Element {
    const { numerator, denominator, charts } = this.props.fractionState;
    const { equal, maxCharts } = this.props;

    // Build an array of whole pie charts, empty if value of fraction is not greater than one
    const wholePies = new Array<JSX.Element>();
      for(let i = 1; i < charts; i++) {
        // add "whole" charts as necessary for fractions > 1
        wholePies.push(<PieFraction equal={equal} numerator={denominator} denominator={denominator}/>);
      }

     // The last pie chart, the fractional part
     const remainder = (numerator === 0) ? <PieFraction equal={equal} numerator={0} denominator={denominator}/> : 
                                         <PieFraction equal={equal} numerator={numerator - (charts - 1) * denominator} denominator={denominator}/>
    return (
       <div className={styles.chartContainer} style={this.chartWidth(charts, maxCharts)}>
            {wholePies}
            {remainder}
       </div>
    );
  }
}
