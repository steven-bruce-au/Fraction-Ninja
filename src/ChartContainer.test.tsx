import React from 'react';
import * as renderer from 'react-test-renderer';
import ChartContainer from './ChartContainer';
import { FractionState } from './FractionState';
import util from 'util';


class EmptyParent implements renderer.TestRendererOptions {

   // override default implementation
   createNodeMock(element: ReactElement): any {
    const doc = document.implementation.createHTMLDocument();
    return { parentElement: doc.body };
   }
}

it('render chart container with two whole charts and a 1/3 pie chart in green', () => {

  const fractionState = new FractionState(1,3);

  const component = renderer.create(<ChartContainer equal='true' maxCharts='3' fractionState='${fractionState}'/> , new EmptyParent());

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();


});

