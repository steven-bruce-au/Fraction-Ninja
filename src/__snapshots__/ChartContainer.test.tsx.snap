// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render chart container with two whole charts and a 1/3 pie chart in green 1`] = `
<div
  className="chartContainer"
  style={
    Object {
      "width": "NaN%",
    }
  }
>
  <div
    className="chart"
  >
    <div
      className="recharts-responsive-container"
      style={
        Object {
          "height": "100%",
          "maxHeight": undefined,
          "minHeight": undefined,
          "minWidth": undefined,
          "width": "100%",
        }
      }
    >
      <div
        style={
          Object {
            "display": "none",
            "height": 0,
            "position": "absolute",
            "visibility": "hidden",
            "width": 0,
          }
        }
      />
    </div>
  </div>
</div>
`;
